<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <!-- jQuery -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.10.3/jquery-ui.min.js"></script>
  
  <!-- bootstrap -->
  <!--<link href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" rel="stylesheet"  id="style">-->
  <script src="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/js/bootstrap.min.js"></script>
  
  <!-- highlight.js -->
  <!--<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" rel="stylesheet" id="code-style">-->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/highlight.min.js"></script>
  <script>
  hljs.LANGUAGES.r=function(a){var b="([a-zA-Z]|\\.[a-zA-Z.])[a-zA-Z0-9._]*";return{c:[a.HCM,{b:b,l:b,k:{keyword:"function if in break next repeat else for return switch while try tryCatch|10 stop warning require library attach detach source setMethod setGeneric setGroupGeneric setClass ...|10",literal:"NULL NA TRUE FALSE T F Inf NaN NA_integer_|10 NA_real_|10 NA_character_|10 NA_complex_|10"},r:0},{cN:"number",b:"0[xX][0-9a-fA-F]+[Li]?\\b",r:0},{cN:"number",b:"\\d+(?:[eE][+\\-]?\\d*)?L\\b",r:0},{cN:"number",b:"\\d+\\.(?!\\d)(?:i\\b)?",r:0},{cN:"number",b:"\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[a.BE],r:0},{cN:"string",b:"'",e:"'",c:[a.BE],r:0}]}}(hljs); </script>
  <!--<script type="text/javascript", src="https://yandex.st/highlightjs/7.3/languages/r.min.js"></script>-->
  
  <!-- Manific Popup -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.8.9/jquery.magnific-popup.min.js"></script>
  
  <script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script>
  <script defer="defer">
  // Function to generate the dynamic table of contents
  jQuery.fn.generate_TOC = function () {
    var base = $(this[0]);
  
    var selectors = ['h1', 'h2', 'h3', 'h4'];
  
    var last_ptr = [{}, {}, {}, {}];
  
    var anchors = {};
  
    generate_anchor = function (text) {
      var test = text.replace(/\W/g, '_');
  
      while(test in anchors){
        //if no suffix, add one
        if(test.match(/_\d+$/) === null){
          test = test + "_2";
        }
        //else generate unique id for duplicates by adding one to the suffix
        else {
          test = test.replace(/_(\d+)$/, function(match, number){ var num=+number+1; return("_" + num) });
        }
      }
      anchors[test]=1;
      return(test);
    }
  
    $(selectors.join(',')).filter(function(index) { return $(this).parent().attr("id") != 'header'; }).each(function () {
  
      var heading = $(this);
      var idx = selectors.indexOf(heading.prop('tagName').toLowerCase());
      var itr = 0;
  
      while (itr <= idx) {
        if (jQuery.isEmptyObject(last_ptr[itr])) {
          last_ptr[itr] = $('<ul>').addClass('nav');
          if (itr === 0) {
            base.append(last_ptr[itr])
          } else {
            if(last_ptr[itr-1].children('li').length === 0){
              last_ptr[itr-1].append(last_ptr[itr]);
            }
            else {
              last_ptr[itr - 1].children('li').last().append(last_ptr[itr]);
            }
          }
        }
        itr++;
      }
      var anchor = generate_anchor(heading.text());
      heading.attr('id', anchor);
      var a = $('<a>')
      .text(heading.text())
      .attr('href', '#' + anchor);
  
    var li = $('<li>')
      .append(a);
  
    last_ptr[idx].append(li);
    for (i = idx + 1; i < last_ptr.length; i++) {
      last_ptr[i] = {};
    }
    });
  }
  /* run scripts when document is ready */
  $(function() {
    "use strict";
  
    var $window = $(window);
    var $body = $(document.body);
  
    /* size of thumbnails */
  
    var hidden_types = ['source']
    var output_types = ['output', 'message', 'warning', 'error']
  
    /* style tables */
    $('table').addClass('table table-striped table-bordered table-hover table-condensed');
  
    $('pre code').each(function(i, e) {
      hljs.highlightBlock(e);
    });
  
    /* Magnific Popup */
    $(".thumbnail").each(function(){
      $(this).magnificPopup({
        disableOn: 768,
        closeOnContentClick: true,
  
        type: 'image',
        items: {
          src: $(this).find('img').attr('src'),
        }
      });
    });
  
    function toggle_block(obj, show) {
      var span = obj.find('span');
      if(show === true){
        span.removeClass('glyphicon-chevron-up').addClass('glyphicon-chevron-down');
        obj.next('pre').slideDown();
      }
      else {
        span.removeClass('glyphicon-chevron-down').addClass('glyphicon-chevron-up');
        obj.next('pre').slideUp();
      }
    }
  
    function toggle_thumbnails(imgs, show){
      if(show === true){
        imgs.parents().show()
        imgs.slideDown();
      }
      else {
        imgs.slideUp(400, function(){ $(this).parent().hide(); });
      }
    }
  
    function global_toggle(obj){
      var type = obj.attr('type');
      var show = !obj.parent('li').hasClass('active');
      if(show === true){
        obj.parent('li').addClass('active');
      }
      else{
        obj.parent('li').removeClass('active');
      }
      if(type == 'figure'){
        toggle_thumbnails($('.thumbnail img'), show);
      }
      else {
        $('.toggle.' + type).each(function() { toggle_block($(this), show); });
      }
    }
  
    /* onclick toggle next code block */
    $('.toggle').click(function() {
      var span = $(this).find('span');
      toggle_block($(this), !span.hasClass('glyphicon-chevron-down'));
      return false
    })
  
    // global toggles
    $('.toggle-global').click(function(){
      var type = $(this).attr('type');
      if(type === 'all-source'){
          $('li a.source').each(function() {
            global_toggle($(this));
          });
        }
      else if(type === 'all-output'){
        $.each(output_types, function(i, val){
          console.log(val);
          global_toggle($('li a.' + val));
        });
      }
      else {
        console.log($(this));
        global_toggle($(this));
      }
      return false;
    });
    /* table of contents */
    if($(['h1', 'h2', 'h3', 'h4'].join(',')).length > 0){
      $('body > #wrap > .container > .row').append('<div class="col-md-2"><div id="toc" class="well sidebar sidenav affix hidden-print"/></div>');
      $('#toc').generate_TOC();
    }
  
    $.each(hidden_types, function(i, type) {
      $('li[type=' + type + ']').each(function(){ global_toggle($(this)); });
    });
  
    /* remove paragraphs with no content */
    $('p:empty').remove();
  
    $body.scrollspy({
      target: '.sidebar',
    });
  
    /* theme switch */
    $('.theme-switch').click(function(){
      var css = $('link[title=' + $(this).attr('title') + ']');
      $('#theme[rel=stylesheet]').attr('href', css.attr('href'));
      $('.theme-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
    /* code style switch */ //TODO use same function for both of these?
    $('.highlight-switch').click(function(){
      var css = $('link[title="' + $(this).attr('title') + '"]');
      $('#highlight[rel=stylesheet]').attr('href', css.attr('href'));
      $('.highlight-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
  
    //TODO refresh on show/hide
    $window.on('load', function () {
      $body.scrollspy('refresh');
    })
  
  });
  
  </script>
  <style>
  /* Knitr_bootstrap styles */
  #header {
    display: none !important;
    visibility: hidden !important;
  }
  #wrap .container-fluid {
    padding: 0;
    overflow: hidden;
  }
  .toggle{
    text-transform: capitalize;
  }
  
  .toggle-global{
    text-transform: capitalize;
  }
  
  /* Sticky footer styles */
  * {
    margin:0;
  }
  html,
  body {
      height: 100%;
      padding:0 !important;
      /* The html and body elements cannot have any padding or margin. */
      /*overflow-x: hidden;*/
  }
  
  /* Wrapper for page content to push down footer */
  #wrap {
      min-height: 100%;
      height: auto !important;
      height: 100%;
      /* Negative indent footer by it's height */
      margin: 0 auto -120px;
  }
  
  /* Set the fixed height of the footer here */
  #push,
  #footer {
      height: 120px;
  }
  
  #footer {
    text-align: center;
  }
  
  /* Top level subheader elements.  These are the first nested items underneath a header element. */
  .header li {
    font-size: 20px;
  }
  
  /* Makes the font smaller for all subheader elements. */
  .sub-header li {
      font-size: 12px;
  }
  
  button.thumbnails {
    margin-left:0px;
  }
  
  /*
   * Side navigation
   *
   * Scrollspy and affixed enhanced navigation to highlight sections and secondary
   * sections of docs content.
   */
  
  /* By default it's not affixed in mobile views, so undo that */
  .sidebar.affix {
    position: static;
  }
  
  /* First level of nav */
  .sidenav {
    margin-top: 30px;
    margin-bottom: 30px;
    padding-top:    10px;
    padding-bottom: 10px;
    border-radius: 5px;
  }
  
  /* All levels of nav */
  .sidebar .nav > li > a {
    display: block;
    padding: 5px 20px;
  }
  .sidebar .nav > li > a:hover,
  .sidebar .nav > li > a:focus {
    text-decoration: none;
    border-right: 1px solid;
  }
  .sidebar .nav > .active > a,
  .sidebar .nav > .active:hover > a,
  .sidebar .nav > .active:focus > a {
    font-weight: bold;
    background-color: transparent;
    border-right: 1px solid;
  }
  
  /* Nav: second level (shown on .active) */
  .sidebar .nav .nav {
    display: none; /* Hide by default, but at >768px, show it */
    margin-bottom: 8px;
  }
  .sidebar .nav .nav > li > a {
    padding-top:    3px;
    padding-bottom: 3px;
    padding-left: 30px;
    font-size: 90%;
  }
  
  .sidebar .nav .nav .nav > li > a {
    padding-left: 40px;
  }
  .sidebar .nav .nav .nav .nav > li > a {
    padding-left: 50px;
  }
  
  /* Show and affix the side nav when space allows it */
  @media screen and (min-width: 992px) {
    .sidebar .nav > .active > ul {
      display: block;
    }
    /* Widen the fixed sidebar */
    .sidebar.affix,
    .sidebar.affix-bottom {
      width: 213px;
    }
    .sidebar.affix-top,
    .sidebar.affix {
      position: fixed; /* Undo the static from mobile first approach */
      top: 30px;
    }
    .sidebar.affix-bottom {
      position: absolute; /* Undo the static from mobile first approach */
    }
    .sidebar.affix-bottom .sidenav,
    .sidebar.affix .sidenav {
      margin-top: 0;
      margin-bottom: 0;
    }
  }
  @media screen and (min-width: 1200px) {
    /* Widen the fixed sidebar again */
    .sidebar.affix-bottom,
    .sidebar.affix {
      width: 263px;
    }
  }
  
  #toc {
    padding: 10px 0px;
    margin:0;
    border:0;
  }
  
  
  .panel pre {
    margin: 0;
    padding: 0;
    border: 0;
  }
  button + pre {
    margin: 0;
    padding: 0;
  }
  pre code {
    border-radius: 0;
  }
  /* Magnific Popup CSS */
  .mfp-bg {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1042;
    overflow: hidden;
    position: fixed;
    background: #0b0b0b;
    opacity: 0.8;
    filter: alpha(opacity=80); }
  
  .mfp-wrap {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1043;
    position: fixed;
    outline: none !important;
    -webkit-backface-visibility: hidden; }
  
  .mfp-container {
    text-align: center;
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    padding: 0 8px;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box; }
  
  .mfp-container:before {
    content: '';
    display: inline-block;
    height: 100%;
    vertical-align: middle; }
  
  .mfp-align-top .mfp-container:before {
    display: none; }
  
  .mfp-content {
    position: relative;
    display: inline-block;
    vertical-align: middle;
    margin: 0 auto;
    text-align: left;
    z-index: 1045; }
  
  .mfp-inline-holder .mfp-content,
  .mfp-ajax-holder .mfp-content {
    width: 100%;
    cursor: auto; }
  
  .mfp-ajax-cur {
    cursor: progress; }
  
  .mfp-zoom-out-cur,
  .mfp-zoom-out-cur .mfp-image-holder .mfp-close {
    cursor: -moz-zoom-out;
    cursor: -webkit-zoom-out;
    cursor: zoom-out; }
  
  .mfp-zoom {
    cursor: pointer;
    cursor: -webkit-zoom-in;
    cursor: -moz-zoom-in;
    cursor: zoom-in; }
  
  .mfp-auto-cursor .mfp-content {
    cursor: auto; }
  
  .mfp-close,
  .mfp-arrow,
  .mfp-preloader,
  .mfp-counter {
    -webkit-user-select: none;
    -moz-user-select: none;
    user-select: none; }
  
  .mfp-loading.mfp-figure {
    display: none; }
  
  .mfp-hide {
    display: none !important; }
  
  .mfp-preloader {
    color: #cccccc;
    position: absolute;
    top: 50%;
    width: auto;
    text-align: center;
    margin-top: -0.8em;
    left: 8px;
    right: 8px;
    z-index: 1044; }
  
  .mfp-preloader a {
    color: #cccccc; }
  
  .mfp-preloader a:hover {
    color: white; }
  
  .mfp-s-ready .mfp-preloader {
    display: none; }
  
  .mfp-s-error .mfp-content {
    display: none; }
  
  button.mfp-close,
  button.mfp-arrow {
    overflow: visible;
    cursor: pointer;
    background: transparent;
    border: 0;
    -webkit-appearance: none;
    display: block;
    padding: 0;
    z-index: 1046;
    -webkit-box-shadow: none;
    box-shadow: none; }
  
  button::-moz-focus-inner {
    padding: 0;
    border: 0; }
  
  .mfp-close {
    width: 44px;
    height: 44px;
    line-height: 44px;
    position: absolute;
    right: 0;
    top: 0;
    text-decoration: none;
    text-align: center;
    opacity: 0.65;
    padding: 0 0 18px 10px;
    color: white;
    font-style: normal;
    font-size: 28px;
    font-family: Arial, Baskerville, monospace; }
    .mfp-close:hover, .mfp-close:focus {
      opacity: 1; }
    .mfp-close:active {
      top: 1px; }
  
  .mfp-close-btn-in .mfp-close {
    color: #333333; }
  
  .mfp-image-holder .mfp-close,
  .mfp-iframe-holder .mfp-close {
    color: white;
    right: -6px;
    text-align: right;
    padding-right: 6px;
    width: 100%; }
  
  .mfp-counter {
    position: absolute;
    top: 0;
    right: 0;
    color: #cccccc;
    font-size: 12px;
    line-height: 18px; }
  
  .mfp-arrow {
    position: absolute;
    opacity: 0.65;
    margin: 0;
    top: 50%;
    margin-top: -55px;
    padding: 0;
    width: 90px;
    height: 110px;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); }
  
  .mfp-arrow:active {
    margin-top: -54px; }
  
  .mfp-arrow:hover,
  .mfp-arrow:focus {
    opacity: 1; }
  
  .mfp-arrow:before, .mfp-arrow:after,
  .mfp-arrow .mfp-b,
  .mfp-arrow .mfp-a {
    content: '';
    display: block;
    width: 0;
    height: 0;
    position: absolute;
    left: 0;
    top: 0;
    margin-top: 35px;
    margin-left: 35px;
    border: medium inset transparent; }
  .mfp-arrow:after,
  .mfp-arrow .mfp-a {
    border-top-width: 13px;
    border-bottom-width: 13px;
    top: 8px; }
  .mfp-arrow:before,
  .mfp-arrow .mfp-b {
    border-top-width: 21px;
    border-bottom-width: 21px; }
  
  .mfp-arrow-left {
    left: 0; }
    .mfp-arrow-left:after,
    .mfp-arrow-left .mfp-a {
      border-right: 17px solid white;
      margin-left: 31px; }
    .mfp-arrow-left:before,
    .mfp-arrow-left .mfp-b {
      margin-left: 25px;
      border-right: 27px solid #3f3f3f; }
  
  .mfp-arrow-right {
    right: 0; }
    .mfp-arrow-right:after,
    .mfp-arrow-right .mfp-a {
      border-left: 17px solid white;
      margin-left: 39px; }
    .mfp-arrow-right:before,
    .mfp-arrow-right .mfp-b {
      border-left: 27px solid #3f3f3f; }
  
  .mfp-iframe-holder {
    padding-top: 40px;
    padding-bottom: 40px; }
  
  .mfp-iframe-holder .mfp-content {
    line-height: 0;
    width: 100%;
    max-width: 900px; }
  
  .mfp-iframe-scaler {
    width: 100%;
    height: 0;
    overflow: hidden;
    padding-top: 56.25%; }
  
  .mfp-iframe-scaler iframe {
    position: absolute;
    display: block;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: black; }
  
  .mfp-iframe-holder .mfp-close {
    top: -40px; }
  
  /* Main image in popup */
  img.mfp-img {
    width: auto;
    max-width: 100%;
    height: auto;
    display: block;
    line-height: 0;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    padding: 40px 0 40px;
    margin: 0 auto; }
  
  /* The shadow behind the image */
  .mfp-figure:after {
    content: '';
    position: absolute;
    left: 0;
    top: 40px;
    bottom: 40px;
    display: block;
    right: 0;
    width: auto;
    height: auto;
    z-index: -1;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: #444444; }
  
  .mfp-figure {
    line-height: 0; }
  
  .mfp-bottom-bar {
    margin-top: -36px;
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    cursor: auto; }
  
  .mfp-title {
    text-align: left;
    line-height: 18px;
    color: #f3f3f3;
    word-wrap: break-word;
    padding-right: 36px; }
  
  .mfp-figure small {
    color: #bdbdbd;
    display: block;
    font-size: 12px;
    line-height: 14px; }
  
  .mfp-image-holder .mfp-content {
    max-width: 100%; }
  
  .mfp-gallery .mfp-image-holder .mfp-figure {
    cursor: pointer; }
  
  @media screen and (max-width: 800px) and (orientation: landscape), screen and (max-height: 300px) {
    /**
     * Remove all paddings around the image on small screen
     */
    .mfp-img-mobile .mfp-image-holder {
      padding-left: 0;
      padding-right: 0; }
  
    .mfp-img-mobile img.mfp-img {
      padding: 0; }
  
    /* The shadow behind the image */
    .mfp-img-mobile .mfp-figure:after {
      top: 0;
      bottom: 0; }
  
    .mfp-img-mobile .mfp-bottom-bar {
      background: rgba(0, 0, 0, 0.6);
      bottom: 0;
      margin: 0;
      top: auto;
      padding: 3px 5px;
      position: fixed;
      -webkit-box-sizing: border-box;
      -moz-box-sizing: border-box;
      box-sizing: border-box; }
  
    .mfp-img-mobile .mfp-bottom-bar:empty {
      padding: 0; }
  
    .mfp-img-mobile .mfp-counter {
      right: 5px;
      top: 3px; }
  
    .mfp-img-mobile .mfp-close {
      top: 0;
      right: 0;
      width: 35px;
      height: 35px;
      line-height: 35px;
      background: rgba(0, 0, 0, 0.6);
      position: fixed;
      text-align: center;
      padding: 0; }
  
    .mfp-img-mobile .mfp-figure small {
      display: inline;
      margin-left: 5px; } }
  @media all and (max-width: 900px) {
    .mfp-arrow {
      -webkit-transform: scale(0.75);
      transform: scale(0.75); }
  
    .mfp-arrow-left {
      -webkit-transform-origin: 0;
      transform-origin: 0; }
  
    .mfp-arrow-right {
      -webkit-transform-origin: 100%;
      transform-origin: 100%; }
  
    .mfp-container {
      padding-left: 6px;
      padding-right: 6px; } }
  .mfp-ie7 .mfp-img {
    padding: 0; }
  .mfp-ie7 .mfp-bottom-bar {
    width: 600px;
    left: 50%;
    margin-left: -300px;
    margin-top: 5px;
    padding-bottom: 5px; }
  .mfp-ie7 .mfp-container {
    padding: 0; }
  .mfp-ie7 .mfp-content {
    padding-top: 44px; }
  .mfp-ie7 .mfp-close {
    top: 0;
    right: 0;
    padding-top: 0; }
  
  //Magnific overrides
  .mfp-image img{
    background: white;
  }
  .mfp-figure:after {
    background: white;
  }
  
  /*
   * Off Canvas navbar toggle right
   * --------------------------------------------------
   */
  
  @media screen and (max-width: 768px) {
    .row-offcanvas .collapsing {
    -webkit-transition: none 0;
      -moz-transition: none 0;
      transition: none 0;
    }
   .row-offcanvas .navbar {
    position: absolute;
    z-index: 2;
      right:0;
      height:100%;
      width:55px;
      border:0;
      background-color:transparent;
    }
    .row-offcanvas .navbar-toggle {
      margin-right: 5px;
      margin-left: 5px;
    }
    .row-offcanvas {
      position: relative;
    }
    .row-offcanvas-right.active .navbar {
    position: absolute;
    z-index: 2;
      right: -28.4%;
      width:40%;
      background-color:#eee;
      border:0 solid #ddd;
      border-left-width:1px;
    }
    .row-offcanvas-right.active {
      left: -30%;
    }
    .row-offcanvas-right.active .navbar-collapse {
      position: relative;
      width: 100%;
    }
    .row-offcanvas .content {
    /*width:calc(100% - 60px);*/
    }
  }
  </style>
</head>
<body>
<div id="wrap">
<div class="container">
<div class="row row-offcanvas row-offcanvas-right">
<div class="contents col-xs-12 col-md-10">
<hr />
title: ‚ÄúQuantitative Big Imaging‚Äù author: ‚ÄúStatistics and Reproducibility‚Äù date: ‚ÄòETHZ: 227-0966-00L‚Äô output: slidy_presentation ‚Äî
<div class="row">
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## Loading required package: knitr
</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## The basic files and libraries needed for most presentations
# creates the libraries and common-functions sections
read_chunk("../common/utility_functions.R")
library(igraph)
library(ICC)
library(grid)</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">require(ggplot2) #for plots
require(lattice) # nicer scatter plots
require(plyr) # for processing data.frames
library(dplyr)
require(grid) # contains the arrow function
require(jpeg)
require(doMC) # for parallel code
require(png) # for reading png images
require(gridExtra)

require(reshape2) # for the melt function
#if (!require("biOps")) {
#  # for basic image processing
#  devtools::install_github("cran/biOps")
#  library("biOps")
#}
## To install EBImage
if (!require("EBImage")) { # for more image processing
  source("http://bioconductor.org/biocLite.R")
  biocLite("EBImage")
}

used.libraries<-c("ggplot2","lattice","plyr","reshape2","grid","gridExtra","biOps","png","EBImage")</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r"># start parallel environment
registerDoMC()
# functions for converting images back and forth
im.to.df<-function(in.img,out.col="val") {
  out.im<-expand.grid(x=1:nrow(in.img),y=1:ncol(in.img))
  out.im[,out.col]<-as.vector(in.img)
  out.im
}
df.to.im<-function(in.df,val.col="val",inv=F) {
  in.vals<-in.df[[val.col]]
  if(class(in.vals[1])=="logical") in.vals<-as.integer(in.vals*255)
  if(inv) in.vals<-255-in.vals
  out.mat<-matrix(in.vals,nrow=length(unique(in.df$x)),byrow=F)
  attr(out.mat,"type")<-"grey"
  out.mat
}
ddply.cutcols<-function(...,cols=1) {
  # run standard ddply command 
  cur.table<-ddply(...)
  cutlabel.fixer<-function(oVal) {
    sapply(oVal,function(x) {
      cnv<-as.character(x)
      mean(as.numeric(strsplit(substr(cnv,2,nchar(cnv)-1),",")[[1]]))
    })
  }
  cutname.fixer<-function(c.str) {
    s.str<-strsplit(c.str,"(",fixed=T)[[1]]
    t.str<-strsplit(paste(s.str[c(2:length(s.str))],collapse="("),",")[[1]]
    paste(t.str[c(1:length(t.str)-1)],collapse=",")
  }
  for(i in c(1:cols)) {
    cur.table[,i]<-cutlabel.fixer(cur.table[,i])
    names(cur.table)[i]<-cutname.fixer(names(cur.table)[i])
  }
  cur.table
}

# since biOps isn't available use a EBImage based version
imgSobel<-function(img) {
  kern<-makeBrush(3)*(-1)
  kern[2,2]<-8
  filter2(img,kern)
}


show.pngs.as.grid<-function(file.list,title.fun,zoom=1) {
  preparePng<-function(x) rasterGrob(readPNG(x,native=T,info=T),width=unit(zoom,"npc"),interp=F)
  labelPng<-function(x,title="junk") (qplot(1:300, 1:300, geom="blank",xlab=NULL,ylab=NULL,asp=1)+
                                        annotation_custom(preparePng(x))+
                                        labs(title=title)+theme_bw(24)+
                                        theme(axis.text.x = element_blank(),
                                              axis.text.y = element_blank()))
  imgList<-llply(file.list,function(x) labelPng(x,title.fun(x)) )
  do.call(grid.arrange,imgList)
}
## Standard image processing tools which I use for visualizing the examples in the script
commean.fun<-function(in.df) {
  ddply(in.df,.(val), function(c.cell) {
    weight.sum<-sum(c.cell$weight)
    data.frame(xv=mean(c.cell$x),
               yv=mean(c.cell$y),
               xm=with(c.cell,sum(x*weight)/weight.sum),
               ym=with(c.cell,sum(y*weight)/weight.sum)
    )
  })
}

colMeans.df<-function(x,...) as.data.frame(t(colMeans(x,...)))

pca.fun<-function(in.df) {
  ddply(in.df,.(val), function(c.cell) {
    c.cell.cov<-cov(c.cell[,c("x","y")])
    c.cell.eigen<-eigen(c.cell.cov)
    
    c.cell.mean<-colMeans.df(c.cell[,c("x","y")])
    out.df<-cbind(c.cell.mean,
                  data.frame(vx=c.cell.eigen$vectors[1,],
                             vy=c.cell.eigen$vectors[2,],
                             vw=sqrt(c.cell.eigen$values),
                             th.off=atan2(c.cell.eigen$vectors[2,],c.cell.eigen$vectors[1,]))
    )
  })
}
vec.to.ellipse<-function(pca.df) {
  ddply(pca.df,.(val),function(cur.pca) {
    # assume there are two vectors now
    create.ellipse.points(x.off=cur.pca[1,"x"],y.off=cur.pca[1,"y"],
                          b=sqrt(5)*cur.pca[1,"vw"],a=sqrt(5)*cur.pca[2,"vw"],
                          th.off=pi/2-atan2(cur.pca[1,"vy"],cur.pca[1,"vx"]),
                          x.cent=cur.pca[1,"x"],y.cent=cur.pca[1,"y"])
  })
}

# test function for ellipse generation
# ggplot(ldply(seq(-pi,pi,length.out=100),function(th) create.ellipse.points(a=1,b=2,th.off=th,th.val=th)),aes(x=x,y=y))+geom_path()+facet_wrap(~th.val)+coord_equal()
create.ellipse.points<-function(x.off=0,y.off=0,a=1,b=NULL,th.off=0,th.max=2*pi,pts=36,...) {
  if (is.null(b)) b<-a
  th<-seq(0,th.max,length.out=pts)
  data.frame(x=a*cos(th.off)*cos(th)+b*sin(th.off)*sin(th)+x.off,
             y=-1*a*sin(th.off)*cos(th)+b*cos(th.off)*sin(th)+y.off,
             id=as.factor(paste(x.off,y.off,a,b,th.off,pts,sep=":")),...)
}
deform.ellipse.draw<-function(c.box) {
  create.ellipse.points(x.off=c.box$x[1],
                        y.off=c.box$y[1],
                        a=c.box$a[1],
                        b=c.box$b[1],
                        th.off=c.box$th[1],
                        col=c.box$col[1])                    
}
bbox.fun<-function(in.df) {
  ddply(in.df,.(val), function(c.cell) {
    c.cell.mean<-colMeans.df(c.cell[,c("x","y")])
    xmn<-emin(c.cell$x)
    xmx<-emax(c.cell$x)
    ymn<-emin(c.cell$y)
    ymx<-emax(c.cell$y)
    out.df<-cbind(c.cell.mean,
                  data.frame(xi=c(xmn,xmn,xmx,xmx,xmn),
                             yi=c(ymn,ymx,ymx,ymn,ymn),
                             xw=xmx-xmn,
                             yw=ymx-ymn
                  ))
  })
}

# since the edge of the pixel is 0.5 away from the middle of the pixel
emin<-function(...) min(...)-0.5
emax<-function(...) max(...)+0.5
extents.fun<-function(in.df) {
  ddply(in.df,.(val), function(c.cell) {
    c.cell.mean<-colMeans.df(c.cell[,c("x","y")])
    out.df<-cbind(c.cell.mean,data.frame(xmin=c(c.cell.mean$x,emin(c.cell$x)),
                                         xmax=c(c.cell.mean$x,emax(c.cell$x)),
                                         ymin=c(emin(c.cell$y),c.cell.mean$y),
                                         ymax=c(emax(c.cell$y),c.cell.mean$y)))
  })
}

common.image.path<-"../common"
qbi.file<-function(file.name) file.path(common.image.path,"figures",file.name)
qbi.data<-function(file.name) file.path(common.image.path,"data",file.name)

th_fillmap.fn<-function(max.val) scale_fill_gradientn(colours=rainbow(10),limits=c(0,max.val))</code></pre>
</div>
<h1 id="course-outline">Course Outline</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">source('../common/schedule.R')</code></pre>
<ul>
<li>25th February - Introduction and Workflows</li>
<li>3rd March - Image Enhancement (A. Kaestner)</li>
<li>10th March - Basic Segmentation, Discrete Binary Structures</li>
<li>17th March - Advanced Segmentation</li>
<li>24th March - Analyzing Single Objects</li>
<li>7th April - Analyzing Complex Objects</li>
<li>14th April - Many Objects and Distributions</li>
<li>21st April - Statistics and Reproducibility</li>
<li>28th April - Dynamic Experiments</li>
<li>12th May - Scaling Up / Big Data</li>
<li>19th May - Guest Lecture - High Content Screening</li>
<li>26th May - Guest Lecture - Machine Learning / Deep Learning and More Advanced Approaches</li>
<li>2nd June - Project Presentations
</div></li>
</ul>
<h1 id="literature-useful-references">Literature / Useful References</h1>
<h3 id="books">Books</h3>
<ul>
<li>Jean Claude, Morphometry with R</li>
<li><a href="http://link.springer.com/book/10.1007%2F978-0-387-77789-4">Online</a> through ETHZ</li>
<li><strong>Chapter 3</strong></li>
<li><a href="http://www.amazon.com/Morphometrics-R-Use-Julien-Claude/dp/038777789X">Buy it</a></li>
<li>John C. Russ, ‚ÄúThe Image Processing Handbook‚Äù,(Boca Raton, CRC Press)</li>
<li>Available <a href="http://dx.doi.org/10.1201/9780203881095">online</a> within domain ethz.ch (or proxy.ethz.ch / public VPN)</li>
<li><a href="http://www.sagepub.com/upm-data/40007_Chapter8.pdf">Hypothesis Testing Chapter</a></li>
<li>Grammar of Graphics: Leland and Wilkinson - <a href="http://www.springer.com/gp/book/9780387245447" class="uri">http://www.springer.com/gp/book/9780387245447</a></li>
</ul>
<h3 id="papers-sites">Papers / Sites</h3>
<ul>
<li>Google/Stanford Statistics Intro</li>
<li><a href="https://www.youtube.com/watch?v=YFC2KUmEebc" class="uri">https://www.youtube.com/watch?v=YFC2KUmEebc</a></li>
<li>MCB 140 P-value lecture at UC Berkeley (Audio)</li>
<li><a href="https://itunes.apple.com/us/itunes-u/mcb-140-fall-2007-general/id461120088?mt=10" class="uri">https://itunes.apple.com/us/itunes-u/mcb-140-fall-2007-general/id461120088?mt=10</a></li>
<li>Correlation and Causation (Video)</li>
<li><a href="https://www.youtube.com/watch?v=YFC2KUmEebc" class="uri">https://www.youtube.com/watch?v=YFC2KUmEebc</a></li>
<li><a href="http://www.mathworks.ch/ch/help/matlab/matlab-unit-test-framework.html">Matlab Unit Testing Documentation</a></li>
<li><a href="http://swcarpentry.github.io/sql-novice-survey/">Databases Introduction</a></li>
<li><a href="http://circos.ca/documentation/course/visualizing-genomic-data.pdf">Visualizing Genomic Data</a> (General Visualization Techniques)</li>
<li><p><a href="http://www.messagelab.monash.edu.au/nimrod">NIMRod Parameter Studies</a></p></li>
<li><p>M.E. Wolak, D.J. Fairbairn, Y.R. Paulsen (2012) Guidelines for Estimating Repeatability. Methods in Ecology and Evolution 3(1):129-137.</p></li>
</ul>
<h1 id="previously-on-qbi">Previously on QBI ‚Ä¶</h1>
<ul>
<li>Image Enhancment</li>
<li>Highlighting the contrast of interest in images</li>
<li>Minimizing Noise</li>
<li>Understanding image histograms</li>
<li>Automatic Methods</li>
<li>Component Labeling</li>
<li>Single Shape Analysis</li>
<li>Complicated Shapes</li>
<li>Distribution Analysis</li>
</ul>
<h1 id="quantitative-big-imaging">Quantitative ‚ÄúBig‚Äù Imaging</h1>
<p>The course has covered imaging enough and there have been a few quantitative metrics, but ‚Äúbig‚Äù has not really entered.</p>
<p>What does <strong>big</strong> mean?</p>
<ul>
<li>Not just / even large</li>
<li>it means being ready for <em>big data</em></li>
<li>volume, velocity, variety (3 V‚Äôs)</li>
<li>scalable, fast, easy to customize</li>
</ul>
<hr />
<p>So what is ‚Äúbig‚Äù imaging</p>
<ul>
<li>doing analyses in a disciplined manner</li>
<li>fixed steps</li>
<li>easy to regenerate results</li>
<li>no <em>magic</em></li>
<li>having everything automated</li>
<li>100 samples is as easy as 1 sample</li>
<li>being able to adapt and reuse analyses</li>
<li>one really well working script and modify parameters</li>
<li>different types of cells</li>
<li>different regions</li>
</ul>
<h1 id="objectives">Objectives</h1>
<ol>
<li>Scientific Studies all try to get to a single number</li>
</ol>
<ul>
<li>Make sure this number is describing the structure well (what we have covered before)</li>
<li>Making sure the number is meaningful (<strong>today!</strong>)</li>
</ul>
<ol>
<li>How do we compare the number from different samples and groups?</li>
</ol>
<ul>
<li>Within a sample or same type of samples</li>
<li>Between samples</li>
</ul>
<ol>
<li>How do we compare different processing steps like filter choice, minimum volume, resolution, etc?</li>
<li>How do we evaluate our parameter selection?</li>
<li>How can we ensure our techniques do what they are supposed to do?</li>
<li>How can we visualize so much data? Are there rules?</li>
</ol>
<h1 id="outline">Outline</h1>
<ul>
<li>Motivation (Why and How?)</li>
<li>Scientific Goals</li>
<li>Reproducibility</li>
<li>Statistical metrics and results</li>
<li>Parameterization</li>
<li>Parameter sweep</li>
<li>Sensitivity analysis</li>
<li>Unit Testing</li>
<li>Visualization</li>
</ul>
<h1 id="what-do-we-start-with">What do we start with?</h1>
<p>Going back to our original cell image</p>
<ol>
<li>We have been able to get rid of the noise in the image and find all the cells (lecture 2-4)</li>
<li>We have analyzed the shape of the cells using the shape tensor (lecture 5)</li>
<li>We even separated cells joined together using Watershed (lecture 6)</li>
<li>We have created even more metrics characterizing the distribution (lecture 7)</li>
</ol>
<p>We have at least a few samples (or different regions), large number of metrics and an almost as large number of parameters to <em>tune</em></p>
<h3 id="how-do-we-do-something-meaningful-with-it">How do we do something meaningful with it?</h3>
<h1 id="correlation-and-causation">Correlation and Causation</h1>
<p>One of the most repeated criticisms of scientific work is that correlation and causation are confused.</p>
<ol>
<li>Correlation</li>
</ol>
<ul>
<li>means a statistical relationship</li>
<li>very easy to show (single calculation)</li>
</ul>
<ol>
<li>Causation</li>
</ol>
<ul>
<li>implies there is a mechanism between A and B</li>
<li>very difficult to show (impossible to prove)</li>
</ul>
<h1 id="controlled-and-observational">Controlled and Observational</h1>
<p>There are two broad classes of data and scientific studies.</p>
<h3 id="observational">Observational</h3>
<ul>
<li>Exploring large datasets looking for trends</li>
<li>Population is random</li>
<li>Not always hypothesis driven</li>
<li>Rarely leads to causation</li>
</ul>
<p>We examined 100 people and the ones with blue eyes were on average 10cm taller</p>
<p>In 100 cake samples, we found a 0.9 correlation between cooking time and bubble size</p>
<hr />
<h3 id="controlled">Controlled</h3>
<ul>
<li>Most scientific studies fall into this category</li>
<li>Specifics of the groups are controlled</li>
<li>Can lead to causation</li>
</ul>
<p>We examined 50 mice with gene XYZ off and 50 gene XYZ on and as the foot size increased by 10%</p>
<p>We increased the temperature and the number of pores in the metal increased by 10%</p>
<h1 id="simple-model-magic-weighted-coin">Simple Model: Magic / Weighted Coin</h1>
<p>incremental: true</p>
<p>Since most of the experiments in science are usually specific, noisy, and often very complicated and are not usually good teaching examples</p>
<ul>
<li>Magic / Biased Coin</li>
<li>You buy a <em>magic</em> coin at a shop</li>
<li>How many times do you need to flip it to <em>prove</em> it is not fair?</li>
<li>If I flip it 10 times and another person flips it 10 times, is that the same as 20 flips?</li>
<li>If I flip it 10 times and then multiple the results by 10 is that the same as 100 flips?</li>
<li>If I buy 10 coins and want to know which ones are fair what do I do?</li>
</ul>
<h1 id="simple-model-magic-weighted-coin-1">Simple Model: Magic / Weighted Coin</h1>
<ol>
<li>Each coin represents a stochastic variable <span class="math inline">ùí≥</span> and each flip represents an observation <span class="math inline">ùí≥<sub><em>i</em></sub></span>.</li>
<li>The act of performing a coin flip <span class="math inline">‚Ñ±</span> is an observation <span class="math inline">ùí≥<sub><em>i</em></sub>‚ÄÑ=‚ÄÑ‚Ñ±(ùí≥)</span></li>
</ol>
<p>We normally assume</p>
<ol>
<li>A <em>fair</em> coin has an expected value of <span class="math inline"><em>E</em>(ùí≥)=0.5‚Üí</span> 50% Heads, 50% Tails</li>
<li>An <em>unbiased</em> flip(er) means</li>
</ol>
<ul>
<li>each flip is independent of the others <br /><span class="math display"><em>P</em>(‚Ñ±<sub>1</sub>(ùí≥)*‚Ñ±<sub>2</sub>(ùí≥))‚ÄÑ=‚ÄÑ<em>P</em>(‚Ñ±<sub>1</sub>(ùí≥))‚ÄÖ*‚ÄÖ<em>P</em>(‚Ñ±<sub>2</sub>(ùí≥))</span><br /></li>
<li>the expected value of the flip is the same as that of the coin <br /><span class="math display">$$ E(\prod_{i=0}^\infty \mathcal{F}_i(\mathcal{X})) = E(\mathcal{X}) $$</span><br /></li>
</ul>
<h1 id="simple-model-to-reality">Simple Model to Reality</h1>
<h3 id="coin-flip">Coin Flip</h3>
<ol>
<li>Each flip gives us a small piece of information about the flipper and the coin</li>
<li>More flips provides more information</li>
</ol>
<ul>
<li>Random / Stochastic variations in coin and flipper cancel out</li>
<li>Systematic variations accumulate</li>
</ul>
<hr />
<h3 id="real-experiment">Real Experiment</h3>
<ol>
<li>Each measurement tells us about our sample, out instrument, and our analysis</li>
<li>More measurements provide more information</li>
</ol>
<ul>
<li>Random / Stochastic variations in sample, instrument, and analysis cancel out</li>
<li><em>Normally</em> the analysis has very little to no stochastic variation</li>
<li>Systematic variations accumulate</li>
</ul>
<h1 id="reproducibility">Reproducibility</h1>
<p>A very broad topic with plenty of sub-areas and deeper meanings. We mean two things by reproducibility</p>
<h3 id="analysis">Analysis</h3>
<p>The process of going from images to numbers is detailed in a clear manner that <em>anyone</em>, <em>anywhere</em> could follow and get the exact (within some tolerance) same numbers from your samples</p>
<ul>
<li>No platform dependence</li>
<li>No proprietary or ‚Äúin house‚Äù algorithms</li>
<li>No manual <em>clicking</em>, <em>tweaking</em>, or <em>copying</em></li>
<li>One script to go from image to result</li>
</ul>
<hr />
<h3 id="measurement">Measurement</h3>
<p>Everything for analysis + taking a measurement several times (noise and exact alignment vary each time) does not change the statistics <em>significantly</em></p>
<ul>
<li>No sensitivity to mounting or rotation</li>
<li>No sensitivity to noise</li>
<li>No dependence on exact illumination</li>
</ul>
<h1 id="reproducible-analysis">Reproducible Analysis</h1>
<p>The basis for reproducible scripts and analysis are scripts and macros. Since we will need to perform the same analysis many times to understand how reproducible it is.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ot">IMAGEFILE=$1</span>
<span class="ot">THRESHOLD=</span>130
<span class="kw">matlab</span> -r <span class="st">&quot;inImage=</span><span class="ot">$IMAGEFILE</span><span class="st">; threshImage=inImage&gt;</span><span class="ot">$THRESHOLD</span><span class="st">; analysisScript;&quot;</span></code></pre></div>
<ul>
<li><strong>or</strong> <code>java -jar ij.jar -macro TestMacro.ijm blobs.tif</code></li>
<li><strong>or</strong> <code>Rscript -e &quot;library(plyr);...&quot;</code></li>
</ul>
<h1 id="comparing-groups-intraclass-correlation-coefficient">Comparing Groups: Intraclass Correlation Coefficient</h1>
The intraclass correlation coefficient basically looking at how similar objects within a group are compared to between groups
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">test.data<-ldply(c("Cats","People","ID","Tiger"),function(class.name) {
  data.frame(name=class.name,values=runif(40))
})
ggplot(test.data,aes(x=name,y=values))+
  geom_point()+
  labs(x="Groups",y="Value",title="Low Group Similarity")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-2-1.png" title="plot of chunk unnamed-chunk-2" alt="plot of chunk unnamed-chunk-2" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">test.data<-ldply(c("Cats","People","ID","Tiger"),function(class.name) {
  data.frame(name=class.name,values=nchar(class.name)+runif(40))
})
ggplot(test.data,aes(x=name,y=values))+
  geom_point()+
  labs(x="Groups",y="Value",title="High Group Similarity")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-3-1.png" title="plot of chunk unnamed-chunk-3" alt="plot of chunk unnamed-chunk-3" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="intraclass-correlation-coefficient-definition">Intraclass Correlation Coefficient Definition</h1>
<p><br /><span class="math display">$$ ICC = \frac{S_A^2}{S_A^2+S_W^2} $$</span><br /></p>
<p>where</p>
<ul>
<li><span class="math inline"><em>S</em><sub><em>A</em></sub><sup>2</sup></span> is the variance among groups or classes</li>
<li>Estimate with the standard deviations of the mean values for each group</li>
<li><span class="math inline"><em>S</em><sub><em>W</em></sub><sup>2</sup></span> is the variance within groups or classes.</li>
<li><p>Estimate with the average of standard deviations for each group</p></li>
<li>1 means 100% of the variance is between classes</li>
<li><p>0 means 0% of the variance is between classes</p></li>
</ul>
<h1 id="intraclass-correlation-coefficient-values">Intraclass Correlation Coefficient: Values</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">library("ICC")

test.data<-ldply(c("Cats","People","ID","Tiger"),function(class.name) {
  data.frame(name=class.name,values=runif(40))
})
icVal<-ICCbare(name,values,data=test.data)
ggplot(test.data,aes(x=name,y=values))+
  geom_point()+
  labs(x="Groups",y="Value",title=paste("Low Group Similarity\n ICC:",icVal))+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-4-1.png" title="plot of chunk unnamed-chunk-4" alt="plot of chunk unnamed-chunk-4" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">test.data<-ldply(c("Cats","People","ID","Tiger"),function(class.name) {
  data.frame(name=class.name,values=nchar(class.name)+runif(40))
})
icVal<-ICCbare(name,values,data=test.data)
ggplot(test.data,aes(x=name,y=values))+
  geom_point()+
  labs(x="Groups",y="Value",title=paste("High Group Similarity\n ICC:",icVal))+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-5-1.png" title="plot of chunk unnamed-chunk-5" alt="plot of chunk unnamed-chunk-5" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="intraclass-correlation-coefficient-values-for-coin-flips">Intraclass Correlation Coefficient: Values for Coin-Flips</h1>
We have one biased coin and try to figure out how many flips we need for the ICC to tell the difference to the normal coin
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">library("ICC")
name.list<-c("Coin A","Coin B")
test.data<-ldply(c(1:length(name.list)),function(class.id) {
  data.frame(name=name.list[class.id],values=runif(100,max=1+2*(class.id-1))>0.5)
})
icVal<-ICCbare(name,values,data=test.data)
ggplot(test.data,aes(x=name,y=values))+
  geom_jitter()+
  labs(x="Groups",y="Value",title=paste("100 flips\n ICC:",icVal))+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-6-1.png" title="plot of chunk unnamed-chunk-6" alt="plot of chunk unnamed-chunk-6" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
With many thousands of flips we eventually see a very strong difference but unless it is very strongly biased ICC is a poor indicator for the differences
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">name.list<-c("Coin A","Coin B")
test.data<-ldply(c(1:length(name.list)),function(class.id) {
  data.frame(name=name.list[class.id],values=runif(20000,max=1+2*(class.id-1))>0.5)
})
icVal<-ICCbare(name,values,data=test.data)
ggplot(test.data,aes(x=name,y=values))+
  geom_jitter()+
  labs(x="Groups",y="Value",title=paste("20,000 flips\n ICC:",icVal))+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-7-1.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="comparing-groups-tests">Comparing Groups: Tests</h1>
<p>Once the reproducibility has been measured, it is possible to compare groups. The idea is to make a test to assess the likelihood that two groups are the same given the data</p>
<ol>
<li>List assumptions</li>
<li>Establish a null hypothesis</li>
</ol>
<ul>
<li>Usually both groups are the same</li>
</ul>
<ol>
<li>Calculate the probability of the observations given the truth of the null hypothesis</li>
</ol>
<ul>
<li>Requires knowledge of probability distribution of the data</li>
<li>Modeling can be exceptionally complicated</li>
</ul>
<hr />
<h3 id="loaded-coin">Loaded Coin</h3>
<p>We have 1 coin from a magic shop</p>
<ul>
<li>our assumptions are</li>
<li>we flip and observe flips of coins accurately and independently</li>
<li>the coin is invariant and always has the same expected value</li>
<li>our null hypothesis is the coin is unbiased <span class="math inline"><em>E</em>(ùí≥)=0.5</span></li>
<li>we can calculate the likelihood of a given observation given the number of flips (p-value)</li>
</ul>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">n.flips<-c(1,5,10)
cf.table<-data.frame(No.Flips=n.flips,PAH=paste(round(1000*0.5^n.flips)/10,"%"))
names(cf.table)<-c("Number of Flips","Probability of All Heads Given Null Hypothesis (p-value)")
kable(cf.table)</code></pre>
<table style="width:104%;">
<colgroup>
<col width="23%" />
<col width="80%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Number of Flips</th>
<th align="left">Probability of All Heads Given Null Hypothesis (p-value)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">50 %</td>
</tr>
<tr class="even">
<td align="right">5</td>
<td align="left">3.1 %</td>
</tr>
<tr class="odd">
<td align="right">10</td>
<td align="left">0.1 %</td>
</tr>
</tbody>
</table>
</div>
<p>How good is good enough?</p>
<h1 id="comparing-groups-students-t-distribution">Comparing Groups: Student‚Äôs T Distribution</h1>
<p>Since we do not usually know our distribution very well <em>or</em> have enough samples to create a sufficient probability model</p>
<h3 id="student-t-distribution"><a href="http://en.wikipedia.org/wiki/Student&#39;s_t-distribution">Student T Distribution</a></h3>
<p>We assume the distribution of our stochastic variable is normal (Gaussian) and the t-distribution provides an estimate for the mean of the underlying distribution based on few observations.</p>
<ul>
<li>We estimate the likelihood of our observed values assuming they are coming from random observations of a normal process</li>
</ul>
<hr />
<h3 id="student-t-test">Student T-Test</h3>
<p>Incorporates this distribution and provides an easy method for assessing the likelihood that the two given set of observations are coming from the same underlying process (null hypothesis)</p>
<ul>
<li>Assume unbiased observations</li>
<li>Assume normal distribution</li>
</ul>
<h1 id="multiple-testing-bias">Multiple Testing Bias</h1>
<p>Back to the magic coin, let‚Äôs assume we are trying to publish a paper, we heard a p-value of &lt; 0.05 (5%) was good enough. That means if we get 5 heads we are good!</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">n.flips<-c(1,4,5)
cf.table<-data.frame(No.Flips=n.flips,PAH=paste(round(1000*0.5^n.flips)/10,"%"))
names(cf.table)<-c("Number of Flips","Probability of All Heads Given Null Hypothesis (p-value)")
kable(cf.table)</code></pre>
<table style="width:104%;">
<colgroup>
<col width="23%" />
<col width="80%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Number of Flips</th>
<th align="left">Probability of All Heads Given Null Hypothesis (p-value)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">50 %</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">6.2 %</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="left">3.1 %</td>
</tr>
</tbody>
</table>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">n.friends<-c(1,10,20,40,80)
cfr.table<-data.frame(No.Friends=n.friends,PAH=paste(round((1000*(1-(1-0.5^5)^n.friends)))/10,"%"))
names(cfr.table)<-c("Number of Friends Flipping","Probability Someone Flips 5 heads")
kable(cfr.table)</code></pre>
<table>
<thead>
<tr class="header">
<th align="right">Number of Friends Flipping</th>
<th align="left">Probability Someone Flips 5 heads</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">3.1 %</td>
</tr>
<tr class="even">
<td align="right">10</td>
<td align="left">27.2 %</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="left">47 %</td>
</tr>
<tr class="even">
<td align="right">40</td>
<td align="left">71.9 %</td>
</tr>
<tr class="odd">
<td align="right">80</td>
<td align="left">92.1 %</td>
</tr>
</tbody>
</table>
</div>
<hr />
<p>Clearly this is not the case, otherwise we could keep flipping coins or ask all of our friends to flip until we got 5 heads and publish</p>
<p>The p-value is only meaningful when the experiment matches what we did.</p>
<ul>
<li>We didn‚Äôt say the chance of getting 5 heads ever was &lt; 5%</li>
<li>We said if we have exactly 5 observations and all of them are heads the likelihood that a fair coin produced that result is &lt;5%</li>
</ul>
<p>Many <a href="http://en.wikipedia.org/wiki/Multiple_comparisons_problem">methods</a> to correct, most just involve scaling <span class="math inline"><em>p</em></span>. The likelihood of a sequence of 5 heads in a row if you perform 10 flips is 5x higher.</p>
<h1 id="multiple-testing-bias-experiments">Multiple Testing Bias: Experiments</h1>
<p>This is very bad news for us. We have the ability to quantify all sorts of interesting metrics</p>
<ul>
<li>cell distance to other cells</li>
<li>cell oblateness</li>
<li>cell distribution oblateness</li>
</ul>
<p>So lets throw them all into a magical statistics algorithm and push the <strong>publish</strong> button</p>
<hr />
<p>With our p value of less than 0.05 and a study with 10 samples in each group, how does increasing the number of variables affect our result</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">make.random.data<-function(n.groups=2,n.samples=10,n.vars=1,rand.fun=runif,group.off.fun=function(grp.id) 0) {
  ldply(1:n.groups,function(c.group) {
    data.frame(group=c.group,
               do.call(cbind,llply(1:n.vars,function(c.var) group.off.fun(c.group)+rand.fun(n.samples)))
               )
  })
}
# only works for two groups
all.t.test<-function(in.data) {
  group1<-subset(in.data,group==1)[,-1,drop=F]
  group2<-subset(in.data,group==2)[,-1,drop=F]
  ldply(1:ncol(group1),function(var.id) {
    tres<-t.test(group1[,var.id],group2[,var.id])
    data.frame(var.col=var.id,
              p.val=tres$p.value,
               method=tres$method,
              var.count=ncol(group1),
              sample.count=nrow(in.data))
    }
  )
}
# run the entire analysis several times to get an average
test.random.data<-function(n.test=10,...) {
  ldply(1:n.test,function(c.test) cbind(test.num=c.test,all.t.test(make.random.data(...))))
}</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">var.range<-round(seq(1,60,length.out=15))
test.cnt<-80
sim.data<-ldply(var.range,function(n.vars) test.random.data(test.cnt,n.vars=n.vars))
sig.likelihood<-ddply(sim.data,.(var.count),function(c.tests) {
  data.frame(sig.vars=nrow(subset(c.tests,p.val<=0.05))/length(unique(c.tests$test.num)))
})</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(sig.likelihood,
       aes(x=var.count,y=sig.vars))+
  geom_point()+geom_line()+
  labs(x="Number of Variables in Study",y="Number of Significant \n (P<0.05) Findings")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-11-1.png" title="plot of chunk unnamed-chunk-11" alt="plot of chunk unnamed-chunk-11" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="multiple-testing-bias-correction">Multiple Testing Bias: Correction</h1>
Using the simple correction factor (number of tests performed), we can make the significant findings constant again
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">sig.likelihood.corr<-ddply(sim.data,.(var.count),function(c.tests) {
  data.frame(sig.vars=nrow(subset(c.tests,p.val<=0.05/var.count))/length(unique(c.tests$test.num)))
})
ggplot(sig.likelihood.corr,
       aes(x=var.count,y=sig.vars))+
  geom_point()+geom_line(aes(color="Corrected"))+
  geom_point(data=sig.likelihood)+
  geom_line(data=sig.likelihood,aes(color="Non-Corrected"))+
  geom_hline(yintercept=0.05,color="green",alpha=0.4,size=2)+
  scale_y_sqrt()+
  labs(x="Number of Variables in Study",y="Number of Significant \n (P<0.05) Findings")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-12-1.png" title="plot of chunk unnamed-chunk-12" alt="plot of chunk unnamed-chunk-12" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<p>So no harm done there we just add this correction factor right? Well what if we have exactly one variable with shift of 1.0 standard deviations from the other.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">var.range<-round(seq(10,60,length.out=10))
test.cnt<-100
one.diff.sample<-function(grp.id) ifelse(grp.id==2,.10,0)
sim.data.diff<-ldply(var.range,function(n.samples) 
  test.random.data(test.cnt,n.samples=n.samples,
                   rand.fun=function(n.cnt) rnorm(n.cnt,mean=1,sd=0.1),
                   group.off.fun=one.diff.sample))</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(sim.data.diff,aes(x=sample.count,y=p.val))+
  geom_point()+
  geom_smooth(aes(color=" 1 Variable"))+
  geom_hline(yintercept=0.05,color="green",alpha=0.4,size=2)+
  labs(x="Number of Samples in Study",y="P-Value for a 10% Difference")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-13-1.png" title="plot of chunk unnamed-chunk-13" alt="plot of chunk unnamed-chunk-13" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="multiple-testing-bias-sample-size">Multiple Testing Bias: Sample Size</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">var.range<-c(1,5,10,20,100) # variable count
sim.data.psig<-ldply(var.range,function(c.vcnt) {
  cbind(var.count=c.vcnt,ddply(sim.data.diff,.(sample.count),function(c.sample) 
    data.frame(prob.sig=nrow(subset(c.sample,p.val<=0.05/c.vcnt))/nrow(c.sample))
  ))
})</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(sim.data.psig,aes(x=sample.count,y=100*prob.sig))+
  geom_line(aes(color=as.factor(var.count)),size=2)+
  ylim(0,100)+
  labs(x="Number of Samples in Study",y="Probability of Finding\n Significant Variable (%)",color="Variables")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-14-1.png" title="plot of chunk unnamed-chunk-14" alt="plot of chunk unnamed-chunk-14" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="parameters">Parameters</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">library(igraph)
make.im.proc.chain<-function(root.node="Raw\nImages",filters=c(),filter.parms=c(),
                             segmentation=c(),segmentation.parms=c(),
                             analysis=c(),analysis.parms=c()) {
  node.names<-c("Raw\nImages",
                filter.parms,filters,
                segmentation.parms,segmentation,
                analysis.parms,analysis
                
                )
  
  c.mat<-matrix(0,length(node.names),length(node.names))
  colnames(c.mat)<-node.names
  rownames(c.mat)<-node.names
  
  
  for(cFilt in filters) {
    c.mat["Raw\nImages",cFilt]<-1
    for(cParm in filter.parms) c.mat[cParm,cFilt]<-1
    for(cSeg in segmentation) {
        c.mat[cFilt,cSeg]<-1
        for(cParm in segmentation.parms) c.mat[cParm,cSeg]<-1
        for(cAnal in analysis) {
          c.mat[cSeg,cAnal]<-1
          for(cParm in analysis.parms) c.mat[cParm,cAnal]<-1
        }
      }
    }
  
  
  g<-graph.adjacency(c.mat,mode="directed")
  V(g)$degree <- degree(g)
  V(g)$label <- V(g)$name
  V(g)$color <- "lightblue"
  V(g)["Raw\nImages"]$color<-"lightgreen"
  for(cAnal in analysis) V(g)[cAnal]$color<-"pink"
  V(g)$size<-30
  for(cParam in c(filter.parms,segmentation.parms,analysis.parms)) {
    V(g)[cParam]$color<-"grey"
    V(g)[cParam]$size<-25
  }
  E(g)$width<-2
  g
  }</code></pre>
</div>
How does a standard image processing chain look?
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">g<-make.im.proc.chain(filters=c("Gaussian\nFilter"),
                      filter.parms=c("3x3\nNeighbors","0.5 Sigma"),
                      segmentation=c("Threshold"),
                      segmentation.parms=c("100"),
                      analysis=c("Shape\nAnalysis","Thickness\nAnalysis")
                      )
plot(g)#,layout=layout.circle) #, layout=layout.circle)# layout.fruchterman.reingold)# layout.kamada.kawai) </code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-15-1.png" title="plot of chunk unnamed-chunk-15" alt="plot of chunk unnamed-chunk-15" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<ul>
<li>Green are the images we start with (measurements)</li>
<li>Blue are processing steps</li>
<li>Gray are use input parameters</li>
<li>Pink are the outputs</li>
</ul>
<h1 id="the-full-chain">The Full Chain</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">library(igraph)
g<-make.im.proc.chain(filters=c("Gaussian\nFilter","Median\nFilter","Diffusion\nFilter","No\nFilter",
                                "Laplacian\nFilter"),
                      segmentation=c("Threshold","Hysteresis\nThreshold","Automated"),
                      analysis=c("Shape\nAnalysis","Thickness\nAnalysis","Distribution\nAnalysis",
                                 "Skeleton\nAnalysis","2 Point\nCorr","Curvature")
                      )
plot(g,layout=layout.reingold.tilford) #, layout=layout.circle)# layout.fruchterman.reingold)# layout.kamada.kawai) </code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-16-1.png" title="plot of chunk unnamed-chunk-16" alt="plot of chunk unnamed-chunk-16" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="the-full-chain-with-parameters">The Full Chain (with Parameters)</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">g<-make.im.proc.chain(filters=c("Gaussian\nFilter","Median\nFilter","Diffusion\nFilter"),
                      filter.parms=c("3x3\nNeighbors","5x5\nNeighbors","7x7\nNeighbors",
                                     "0.5 Sigma","1.0 Sigma","1.2 Sigma"),
                      segmentation=c("Threshold","Hysteresis\nThreshold","Automated"),
                      segmentation.parms=paste(seq(90,110,length.out=3)),
                      analysis=c("Shape\nAnalysis","Thickness\nAnalysis","Distribution\nAnalysis","Skeleton\nAnalysis","2 Point\nCorr")
                      )
plot(g,layout=layout.lgl(g,maxiter=10000,root=1)) #, layout=layout.circle)# layout.fruchterman.reingold)# layout.kamada.kawai) </code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-17-1.png" title="plot of chunk unnamed-chunk-17" alt="plot of chunk unnamed-chunk-17" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<ul>
<li>A <strong>mess</strong>, over 1080 combinations for just one sample (not even exploring a very large range of threshold values)</li>
<li>To calculate this for even one sample can take days (weeks, years)</li>
<li>512 x 512 x 512 foam sample <span class="math inline">‚Üí</span> 12 weeks of processing time</li>
<li>1024 x 1024 x 1024 femur bone <span class="math inline">‚Üí</span> 1.9 years</li>
<li>Not all samples are the same</li>
<li>Once the analysis is run we have a ton of data</li>
<li>femur bone <span class="math inline">‚Üí</span> 60 million shapes analyzed</li>
<li>What do we even want?</li>
<li>How do we judge the different results?</li>
</ul>
<h1 id="qualitative-vs-quantitative">Qualitative vs Quantitative</h1>
<p>Given the complexity of the tree, we need to do some pruning</p>
<h3 id="qualitative-assessment">Qualitative Assessment</h3>
<ul>
<li>Evaluating metrics using visual feedback</li>
<li>Compare with expectations from other independent techniques or approach</li>
<li>Are there artifacts which are included in the output?</li>
<li>Do the shapes look correct?</li>
<li>Are they distributed as expected?</li>
<li>Is their orientation meaningful?</li>
</ul>
<hr />
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">title.fun<-function(file.name) ""
show.pngs.as.grid(Sys.glob("ext-figures/poros*.png"),title.fun,zoom=0.5)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-18-1.png" title="plot of chunk unnamed-chunk-18" alt="plot of chunk unnamed-chunk-18" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="quantitative-metrics">Quantitative Metrics</h1>
<p>With a quantitative approach, we can calculate the specific shape or distribution metrics on the sample with each parameter and establish the relationship between parameter and metric.</p>
<h3 id="parameter-sweep">Parameter Sweep</h3>
<p>The way we do this is usually a parameter sweep which means taking one (or more) parameters and varying them between the reasonable bounds (judged qualitatively).</p>
<hr />
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">source('../common/shapeAnalysisProcess.R')
source('../common/commonReportFunctions.R')
# read and correct the coordinate system
thresh.fun<-function(x) {
  pth<-rev(strsplit(x,"/")[[1]])[2]
  t<-strsplit(pth,"_")[[1]][3]
  as.numeric(substring(t,2,nchar(t)))
}
readfcn<-function(x) cbind(compare.foam.corrected(x,
                                                  checkProj=F
                                                  #force.scale=0.011 # force voxel size to be 11um
                                                  ),
                           thresh=thresh.fun(x) # how to parse the sample names
                           )
# Where are the csv files located
rootDir<-"../common/data/mcastudy" 
clpor.files<-Sys.glob(paste(rootDir,"/a*/lacun_0.csv",sep="/")) # list all of the files

# Read in all of the files
all.lacun<-ldply(clpor.files,readfcn,.parallel=T)</code></pre>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r"> ggplot(all.lacun,aes(y=VOLUME*1e9,x=thresh))+
  geom_jitter(alpha=0.1)+geom_smooth()+
  theme_bw(24)+labs(y="Volume (um3)",x="Threshold Value",color="Threshold")+ylim(0,1000)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-19-1.png" title="plot of chunk unnamed-chunk-19" alt="plot of chunk unnamed-chunk-19" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="is-it-always-the-same">Is it always the same?</h1>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r"> ggplot(subset(all.lacun,thresh %% 1000==0),aes(y=VOLUME*1e9,x=as.factor(thresh)))+
  geom_violin()+
  theme_bw(24)+labs(y="Volume (um3)",x="Threshold Value",color="Threshold")+ylim(0,1000)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-20-1.png" title="plot of chunk unnamed-chunk-20" alt="plot of chunk unnamed-chunk-20" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r"> ggplot(all.lacun,aes(y=PCA1_Z,x=thresh))+
  geom_jitter(alpha=0.1)+geom_smooth()+
  theme_bw(24)+labs(y="Orientation",x="Threshold Value",color="Threshold")</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-21-1.png" title="plot of chunk unnamed-chunk-21" alt="plot of chunk unnamed-chunk-21" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="sensitivity">Sensitivity</h1>
<p>Sensitivity is defined in control system theory as the change in the value of an output against the change in the input. <br /><span class="math display">$$ S = \frac{|\Delta \textrm{Metric}|}{|\Delta \textrm{Parameter}|} $$</span><br /></p>
<p>Such a strict definition is not particularly useful for image processing since a threshold has a unit of intensity and a metric might be volume which has <span class="math inline"><em>m</em><sup>3</sup></span> so the sensitivity becomes volume per intensity.</p>
<hr />
<h3 id="practical-sensitivity">Practical Sensitivity</h3>
<p>A more common approach is to estimate the variation in this parameter between images or within a single image (automatic threshold methods can be useful for this) and define the sensitivity based on this variation. It is also common to normalize it with the mean value so the result is a percentage.</p>
<p><br /><span class="math display">$$ S = \frac{max(\textrm{Metric})-min(\textrm{Metric})}{avg(\textrm{Metric})} $$</span><br /></p>
<h1 id="sensitivity-real-measurements">Sensitivity: Real Measurements</h1>
<p>In this graph it is magnitude of the slope. The steeper the slope the more the metric changes given a small change in the parameter</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">poresum<-function(all.data) ddply(all.data,.(thresh),function(c.sample) {
  data.frame(Count=nrow(c.sample),
             Volume=mean(c.sample$VOLUME*1e9),
             Stretch=mean(c.sample$AISO),
             Oblateness=mean(c.sample$OBLATENESS),
             #Lacuna_Density_mm=1/mean(c.sample$DENSITY_CNT),
             Length=mean(c.sample$PROJ_PCA1*1000),
             Width=mean(c.sample$PROJ_PCA2*1000),
             Height=mean(c.sample$PROJ_PCA3*1000),
             Orientation=mean(abs(c.sample$PCA1_Z)))
})
comb.summary<-cbind(poresum(all.lacun),Phase="Lacuna")
splot<-ggplot(comb.summary,aes(x=thresh))
splot+geom_line(aes(y=Count))+geom_point(aes(y=Count))+scale_y_log10()+
  theme_bw(24)+labs(y="Object Count",x="Threshold",color="Phase")</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-22-1.png" title="plot of chunk unnamed-chunk-22" alt="plot of chunk unnamed-chunk-22" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<p>Comparing Different Variables we see that the best (lowest) value for the count sensitivity is the highest for the volume and anisotropy.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">calc.sens<-function(in.df) {
  data.frame(sens.cnt=100*with(in.df,(max(Count)-min(Count))/mean(Count)),
             sens.vol=100*with(in.df,(max(Volume)-min(Volume))/mean(Volume)),
             sens.stretch=100*with(in.df,(max(Stretch)-min(Stretch))/mean(Stretch))
             )
}
sens.summary<-ddply.cutcols(comb.summary,.(cut_interval(thresh,5)),calc.sens)
ggplot(sens.summary,aes(x=thresh))+
  geom_line(aes(y=sens.cnt,color="Count"))+
  geom_line(aes(y=sens.vol,color="Volume"))+
  geom_line(aes(y=sens.stretch,color="Anisotropy"))+
  labs(x="Threshold",y="Sensitivity (%)",color="Metric")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-23-1.png" title="plot of chunk unnamed-chunk-23" alt="plot of chunk unnamed-chunk-23" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h3 id="which-metric-is-more-important">Which metric is more important?</h3>
<h1 id="unit-testing">Unit Testing</h1>
<p>ÔøºIn computer programming, unit testing is a method by which individual units of source code, sets of one or more computer program modules together with associated control data, usage procedures, and operating procedures, are tested to determine if they are fit for use.</p>
<ul>
<li>Intuitively, one can view a unit as the smallest testable part of an application</li>
<li>Unit testing is possible with every language</li>
<li>Most (Java, C++, Matlab, R, Python) have built in support for automated testing and reporting</li>
</ul>
<h1 id="unit-testing-continued">Unit Testing Continued</h1>
<p>The first requirement for unit testing to work well is to have you tools divided up into small independent parts (functions)</p>
<ul>
<li>Each part can then be tested independently (unit testing)</li>
<li>If the tests are well done, units can be changed and tested independently</li>
<li>Makes upgrading or expanding tools <em>easy</em></li>
<li>The entire path can be tested (integration testing)</li>
<li>Catches mistakes in integration or <em>glue</em></li>
</ul>
<hr />
<p>Ideally with realistic but simulated test data</p>
<ul>
<li>The utility of the testing is only as good as the tests you make</li>
</ul>
<h3 id="example">Example</h3>
<ul>
<li>Given the following function <code>function vxCnt=countVoxs(inImage)</code></li>
<li>We can right the following tests</li>
<li>testEmpty2d</li>
</ul>
<p><code>assert countVoxs(zeros(3,3)) == 0</code></p>
<ul>
<li>testEmpty3d</li>
</ul>
<p><code>assert countVoxs(zeros(3,3,3)) == 0</code></p>
<ul>
<li>testDiag3d</li>
</ul>
<p><code>assert countVoxs(eye(3)) == 3</code></p>
<h1 id="unit-testing-examples">Unit Testing: Examples</h1>
<ul>
<li>Given the following function <code>function shapeTable=shapeAnalysis(inImage)</code> We should decompose the task into sub-components</li>
<li><p><code>function clImage=componentLabel(inImage)</code></p></li>
<li><code>function objInfo=analyzeObject(inObject)</code></li>
<li><code>function vxCnt=countVoxs(inObject)</code></li>
<li><code>function covMat=calculateCOV(inObject)</code></li>
<li><code>function shapeT=calcShapeT(covMat)</code></li>
<li><code>function angle=calcOrientation(shapeT)</code></li>
<li><p><code>function aniso=calcAnisotropy(shapeT)</code></p></li>
</ul>
<h1 id="unit-testing-in-imagej">Unit Testing in ImageJ</h1>
<p><a href="https://github.com/imagej/ij1-tests/blob/master/src/test/java/ij/VirtualStackTest.java">On this page</a></p>
<iframe src="https://raw.githubusercontent.com/imagej/ij1-tests/master/src/test/java/ij/VirtualStackTest.java" width="100%" height="800">
</iframe>
<h1 id="unit-testing-in-knime">Unit Testing in KNIME</h1>
<p><a href="https://tech.knime.org/community/developers">Read more</a> and <a href="https://www.knime.org/files/kos-11/KNIME_Testing.pdf">Here</a> The Java-based unit-testing can be used (JUnit) before any of the plugins are compiled, additionally entire workflows can be made to test the objects using special testing nodes like</p>
<ul>
<li>difference node (check the if two values are different)</li>
</ul>
<p><img src="ext-figures/KnimeTests.png" alt="KNIME Tests" /></p>
<ul>
<li>disturber node (insert random / missing values to determine fault tolerance)</li>
</ul>
<h1 id="test-driven-programming">Test Driven Programming</h1>
<p>Test Driven programming is a style or approach to programming where the tests are written before the functional code. Like very concrete specifications. It is easy to estimate how much time is left since you can automatically see how many of the tests have been passed. You and your collaborators are clear on the utility of the system.</p>
<ol>
<li>shapeAnalysis must give an anisotropy of 0 when we input a sphere</li>
<li>shapeAnalysis must give the center of volume within 0.5 pixels</li>
<li>shapeAnalysis must run on a 1000x1000 image in 30 seconds</li>
</ol>
<h1 id="visualization">Visualization</h1>
<p>One of the biggest problems with <em>big</em> sciences is trying to visualize a lot of heterogeneous data.</p>
<ul>
<li>Tables are difficult to interpret</li>
<li>3D Visualizations are very difficult to compare visually</li>
<li>Contradictory necessity of simple single value results and all of the data to look for trends and find problems</li>
</ul>
<h1 id="bad-graphs">Bad Graphs</h1>
<p>There are too many graphs which say</p>
<ul>
<li>‚Äòmy data is very complicated‚Äô</li>
<li>‚ÄòI know how to use __ toolbox in Matlab/R/Mathematica‚Äô</li>
</ul>
<p><img src="ext-figures/badImage1.png" alt="3d Plots" /> <img src="ext-figures/badPlot4.png" alt="Spectroscopy" /></p>
<hr />
<ul>
<li>Most programs by default make poor plots</li>
<li>Good visualizations takes time <img src="ext-figures/badImage3.png" alt="Linkage" /></li>
</ul>
<p><img src="ext-figures/badImage2.png" alt="Linkage 2" /></p>
<h1 id="key-ideas">Key Ideas</h1>
<ol>
<li>What is my message?</li>
<li>Does the graphic communicate it clearly?</li>
<li>Is a graphic representation really necessary?</li>
</ol>
<ul>
<li></li>
</ul>
<ol>
<li>Does every line / color serve a purpose?</li>
</ol>
<ul>
<li>Pretend ink is very expensive</li>
</ul>
<hr />
<h3 id="simple-rules">Simple Rules</h3>
<ol>
<li>Never use 3D graphics when it can be avoided (unless you want to be deliberately misleading), our visual system is not well suited for comparing heights of different <img src="ext-figures/3dplot.png" alt="Dumb 3d" /></li>
<li>Pie charts can also be hard to interpret</li>
<li>Background color should almost always be white (not light gray)</li>
<li>Use color palettes adapted to human visual sensitivity</li>
</ol>
<h1 id="what-is-my-message">What is my message</h1>
<ul>
<li>Plots to ‚Äúshow the results‚Äù or ‚Äúget a feeling‚Äù are usually not good</li>
</ul>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">xd<-runif(80)
test.data<-data.frame(x=xd,y=xd+runif(80),z=runif(80))
plot(test.data)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-24-1.png" title="plot of chunk unnamed-chunk-24" alt="plot of chunk unnamed-chunk-24" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<hr />
<ul>
<li>Focus on a single, simple message</li>
<li>X is a little bit correlated with Y
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(test.data,aes(x,y))+
  geom_point()+geom_smooth(method="lm")+
  coord_equal()+
  labs(title="X is weakly correlated with Y")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-25-1.png" title="plot of chunk unnamed-chunk-25" alt="plot of chunk unnamed-chunk-25" style="display: block; margin: auto;" /></a>
</div>
</div>
</div></li>
</ul>
<h1 id="does-my-graphic-communicate-it-clearly">Does my graphic communicate it clearly?</h1>
<ul>
<li>Too much data makes it very difficult to derive a clear message
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">xd<-runif(5000)
test.data<-data.frame(x=xd,y=(xd-0.5)*runif(5000))
ggplot(test.data,aes(x,y))+
  geom_point()+
  coord_equal()+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-26-1.png" title="plot of chunk unnamed-chunk-26" alt="plot of chunk unnamed-chunk-26" style="display: block; margin: auto;" /></a>
</div>
</div>
</div></li>
</ul>
<hr />
<ul>
<li>Filter and reduce information until it is extremely simple</li>
</ul>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(test.data,aes(x,y))+
  stat_binhex(bins=20)+
  geom_smooth(method="lm",aes(color="Fit"))+
  coord_equal()+
  theme_bw(20)+guides(color=F)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-27-1.png" title="plot of chunk unnamed-chunk-27" alt="plot of chunk unnamed-chunk-27" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">ggplot(test.data,aes(x,y))+
  geom_density2d(aes(color="Contour"))+
  geom_smooth(method="lm",aes(color="Linear Fit"))+
  coord_equal()+
  labs(color="Type")+
  theme_bw(20)</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="print_figures/unnamed-chunk-28-1.png" title="plot of chunk unnamed-chunk-28" alt="plot of chunk unnamed-chunk-28" style="display: block; margin: auto;" /></a>
</div>
</div>
</div>
<h1 id="grammar-of-graphics">Grammar of Graphics</h1>
<ul>
<li>What is a grammar?</li>
<li>Set of rules for constructing and validating a sentence</li>
<li>Specifies the relationship and order between the words constituting the sentence</li>
<li>How does this apply to graphics?</li>
<li>If we develop a consistent way of expressing graphics (sentences) in terms of elements (words) we can compose and decompose graphics easily</li>
</ul>
<hr />
<ul>
<li><p>The most important modern work in graphical grammars is ‚ÄúThe Grammar of Graphics‚Äù by Wilkinson, Anand, and Grossman (2005). This work built on earlier work by Bertin (1983) and proposed a grammar that can be used to describe and construct a wide range of statistical graphics.</p></li>
<li><p>This can be applied in R using the ggplot2 library (<small>H. Wickham. ggplot2: elegant graphics for data analysis. Springer New York, 2009.</small>)</p></li>
</ul>
<h1 id="grammar-explained">Grammar Explained</h1>
<p>Normally we think of plots in terms of some sort of data which is fed into a plot command that produces a picture</p>
<ul>
<li>In Excel you select a range and plot-type and click ‚ÄúMake‚Äù</li>
<li>In Matlab you run <code>plot(xdata,ydata,color/shape)</code></li>
</ul>
<ol>
<li>These produces entire graphics (sentences) or at least phrases in one go and thus abstract away from the idea of grammar.</li>
<li>If you spoke by finding entire sentences in a book it would be very ineffective, it is much better to build up word by word</li>
</ol>
<hr />
<h3 id="grammar">Grammar</h3>
<p>Separate the graph into its component parts</p>
<ol>
<li>Data Mapping</li>
</ol>
<ul>
<li><span class="math inline"><em>v</em><em>a</em><em>r</em>1‚ÄÑ‚Üí‚ÄÑ<em>x</em></span>, <span class="math inline"><em>v</em><em>a</em><em>r</em>2‚ÄÑ‚Üí‚ÄÑ<em>y</em></span></li>
</ul>
<p><img src="ext-figures/grammarOfGraphics.png" alt="Graph Decomposed" /></p>
<ol>
<li>Points</li>
<li>Axes / Coordinate System</li>
<li>Labels / Annotation</li>
</ol>
<p>Construct graphics by focusing on each portion independently.</p>
<h1 id="wrapping-up">Wrapping up</h1>
<ul>
<li>I am not a statistician</li>
<li>This is not a statistics course</li>
<li>If you have questions or concerns, Both ETHZ and Uni Zurich offer <strong>free</strong> consultation with real statisticians</li>
<li>They are rarely bearers of good news</li>
</ul>
<hr />
<ul>
<li><p>Simulations (even simple ones) are very helpful (see <a href="knime://LOCAL/Exercise%209%20StatsRepro/StatisticalSignificanceHunter">StatisticalSignificanceHunter</a>)</p></li>
<li>Try and understand the tests you are performing
</div>
</div>
<div class="navbar navbar-fixed-bottom navbar-inverse">
<div class="container">
<div class="navbar-header">
<button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-responsive-collapse">
<span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span>
</button>
</div>
<div id="bottom-navbar" class="navbar-collapse collapse navbar-responsive-collapse">
<ul class="nav navbar-nav navbar-right">
<li class="nav">
<p class="navbar-text">
Toggle
</p>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Code <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Languages
</li>
<li class="active">
<a href="#" class="toggle-global source R" type="source.R">R</a>
</li>
<li>
<a href="#" type="all-source" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Output <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Type
</li>
<li class="active">
<a href="#" class="toggle-global message" type="message">message</a>
</li>
<li>
<a href="#" type="all-output" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="active">
<a href="#" type="figure" class="toggle-global">Figures</a>
</li>
</ul>
</div>
</div>
</div>
</div>
<div id="push">
</div>
<div id="footer">
<div class="container">
<p class="text-muted" id="credit">
Styled with <a href="https://github.com/jimhester/knitrBootstrap">knitrBootstrap</a>
</p>
</div>
</div>
<link rel="stylesheet" id="theme" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" media="screen"></link><link rel="stylesheet" id="highlight" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" media="screen"></link>
</div></li>
</ul>
</body>
</html>
